File: frontend/src/app/core/models/extraction.ts
export type TaskType =
  | 'schema_induction'
  | 'instance_unstructured'
  | 'instance_structured'
  | 'business_logic'
  | 'business_logic_topology';

export type TaskStatus = 'pending' | 'running' | 'succeeded' | 'failed';

export interface ExtractionTaskRead {
  id: string;
  task_type: TaskType;
  status: TaskStatus;
  progress: number;
  output_summary?: Record<string, unknown> | null;
  error_message?: string | null;
  schema_id?: string | null;
  created_at?: string;
}

export interface UnstructuredExtractionRequest {
  schema_id: string;
  file_ids: string[];
  ai_config?: Record<string, unknown> | null;
  model_id?: string | null;
  replace_existing?: boolean;
}

export interface StructuredExtractionRequest {
  schema_id: string;
  mapping_ids: string[];
  replace_existing?: boolean;
}

export interface BusinessLogicExtractionRequest {
  ontology_model_id?: string | null;
  schema_id?: string | null;
  file_ids: string[];
  ai_config?: Record<string, unknown> | null;
  model_id?: string | null;
  schema_version?: number | null;
  type_mapping?: Record<string, string[]> | null;
  name?: string | null;
}

export interface InstanceRead {
  id: string;
  schema_id: string;
  class_id: string | null;
  class_label?: string | null;
  label: string;
  local_name?: string | null;
  source_type: 'ai_unstructured' | 'structured_mapping' | 'manual';
  confidence?: number | null;
  schema_version?: number | null;
  created_at: string;
}

export interface InstanceDataValueWrite {
  property_id: string;
  value: string;
}

export interface InstanceRelationWrite {
  property_id: string;
  object_instance_id: string;
}

export interface InstanceUpdate {
  class_id: string | null;
  data_values: InstanceDataValueWrite[];
  relations: InstanceRelationWrite[];
}

export interface InstanceDetail extends InstanceRead {
  data_values: { property_id: string; property_label: string; value: string }[];
  relations: {
    property_id: string;
    property_label: string;
    object_instance_id: string;
    object_instance_label?: string | null;
    object_label?: string | null;
    object_class_label?: string | null;
    direction?: 'out' | 'in';
  }[];
}

export interface InstanceStat {
  class_id: string;
  class_label: string;
  count: number;
}

export interface InstanceStatsResponse {
  schema_id: string;
  schema_version?: number | null;
  total: number;
  by_class: InstanceStat[];
}

export interface InstanceInventory {
  schema_id: string;
  schema_name: string;
  schema_version: number;
  filter_version?: number | null;
  versions: number[];
  total: number;
  by_class: InstanceStat[];
  recent_tasks: ExtractionTaskRead[];
}

export interface ClearInstancesResult {
  deleted: number;
  schema_id: string;
  schema_version?: number | null;
}

export interface BusinessLogicRuleRead {
  id: string;
  schema_id: string;
  rule_type: 'causality' | 'constraint';
  description: string;
  condition: Record<string, unknown>;
  consequence?: Record<string, unknown> | null;
  action_required?: string | null;
  severity?: string | null;
  created_at: string;
}

File: backend/app/schemas/extraction.py
"""Extraction / instance Pydantic DTOs."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


class ExtractionTaskRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    task_type: Literal[
        "schema_induction",
        "instance_unstructured",
        "instance_structured",
        "business_logic",
        "business_logic_topology",
    ]
    status: Literal["pending", "running", "succeeded", "failed"]
    schema_id: str | None = None
    progress: float
    output_summary: dict | None = None
    error_message: str | None = None
    created_at: datetime | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None


class TaskAccepted(BaseModel):
    task_id: str
    status: Literal["pending"] = "pending"


class UnstructuredExtractionRequest(BaseModel):
    schema_id: str
    file_ids: list[str] = Field(min_length=1)
    ai_config: dict | None = None
    model_id: str | None = None
    # True：抽取前清除本 Schema 当前版本下的非结构化实例，避免重复堆积
    replace_existing: bool = True


class StructuredExtractionRequest(BaseModel):
    schema_id: str
    mapping_ids: list[str] = Field(min_length=1)
    replace_existing: bool = False


class BusinessLogicExtractionRequest(BaseModel):
    ontology_model_id: str | None = None
    schema_id: str | None = None
    file_ids: list[str] = Field(min_length=1)
    ai_config: dict | None = None
    model_id: str | None = None
    schema_version: int | None = None
    type_mapping: dict[str, list[str]] | None = None
    name: str | None = None

    @model_validator(mode="after")
    def _need_target(self) -> BusinessLogicExtractionRequest:
        if not self.ontology_model_id and not self.schema_id:
            raise ValueError("请选择本体模型")
        return self


class ClearInstancesRequest(BaseModel):
    schema_version: int | None = None
    source_types: list[str] | None = None


class ClearInstancesResult(BaseModel):
    deleted: int
    schema_id: str
    schema_version: int | None = None


class InstanceDataValueRead(BaseModel):
    property_id: str
    property_label: str | None = None
    value: str


class InstanceRelationRead(BaseModel):
    property_id: str
    property_label: str | None = None
    object_instance_id: str
    object_instance_label: str | None = None
    object_label: str | None = None
    object_class_label: str | None = None
    direction: str = "out"


class InstanceDataValueWrite(BaseModel):
    property_id: str = Field(min_length=1)
    value: str = Field(min_length=1)


class InstanceRelationWrite(BaseModel):
    property_id: str = Field(min_length=1)
    object_instance_id: str = Field(min_length=1)


class InstanceUpdate(BaseModel):
    """Replace class + data values + outgoing object relations in one save."""

    class_id: str | None = None
    data_values: list[InstanceDataValueWrite] = Field(default_factory=list)
    relations: list[InstanceRelationWrite] = Field(default_factory=list)


class InstanceRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    schema_id: str
    class_id: str | None = None
    class_label: str | None = None
    label: str
    local_name: str | None = None
    source_type: str
    source_ref: dict | None = None
    confidence: float | None = None
    schema_version: int | None = None
    extraction_task_id: str | None = None
    created_at: datetime
    data_values: list[InstanceDataValueRead] = []
    relations: list[InstanceRelationRead] = []


class InstanceStatsItem(BaseModel):
    class_id: str
    class_label: str
    count: int


class InstanceStatsResponse(BaseModel):
    schema_id: str
    schema_version: int | None = None
    total: int
    by_class: list[InstanceStatsItem]


class InstanceInventoryResponse(BaseModel):
    schema_id: str
    schema_name: str
    schema_version: int
    filter_version: int | None = None
    versions: list[int]
    total: int
    by_class: list[InstanceStatsItem]
    recent_tasks: list[ExtractionTaskRead] = []
